python/CMakeLists.txt

```diff
@@ -1,27 +1,163 @@
+
+if (CMAKE_COMPILER_IS_GNUCXX)
+  # Deals with warnings in gcc12 - may not be needed for gcc 13.
+  # To enable this also for clang, make sure to install the 
+  # corresponding llvm-*-dev package (no change needed for the LLVM build).
+  add_definitions(-flto=auto)
+endif()
+
+set(CMAKE_PLATFORM_NO_VERSIONED_SONAME 1)
+
+################################################################################
+# Sources
+################################################################################
+
+declare_mlir_python_sources(CUDAQuantumPythonSources)
+
+declare_mlir_dialect_python_bindings(
+  ADD_TO_PARENT CUDAQuantumPythonSources
+  ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/mlir_cudaq"
+  TD_FILE dialects/QuakeOps.td
+  SOURCES
+    dialects/quake.py
+  DIALECT_NAME quake)
+
+declare_mlir_dialect_python_bindings(
+  ADD_TO_PARENT CUDAQuantumPythonSources
+  ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/mlir_cudaq"
+  TD_FILE dialects/CCOps.td
+  SOURCES
+    dialects/cc.py
+  DIALECT_NAME cc)
+
+declare_mlir_python_extension(CUDAQuantumPythonSources.Extension
+  MODULE_NAME _quakeDialects
+  ADD_TO_PARENT CUDAQuantumPythonSources
+  SOURCES
+   CUDAQuantumExtension.cpp
+   runtime/mlir/py_register_dialects.cpp
+   ../runtime/cudaq/platform/default/rest/PyRemoteRESTQPU.cpp
+   runtime/cudaq/qis/py_qubit_qis.cpp
+   runtime/cudaq/kernels/py_common_kernels.cpp
+   runtime/cudaq/platform/py_alt_launch_kernel.cpp
+   runtime/cudaq/qis/py_execution_manager.cpp 
+   runtime/common/py_ExecutionContext.cpp
+   runtime/cudaq/algorithms/py_sample_async.cpp
+   runtime/cudaq/algorithms/py_observe_async.cpp
+   runtime/cudaq/platform/JITExecutionCache.cpp
+   utils/LinkedLibraryHolder.cpp
+   runtime/cudaq/target/py_runtime_target.cpp
+   runtime/cudaq/target/py_testing_utils.cpp
+   runtime/common/py_SampleResult.cpp
+   runtime/common/py_ObserveResult.cpp
+   runtime/common/py_NoiseModel.cpp
+   runtime/cudaq/spin/py_matrix.cpp
+   runtime/cudaq/spin/py_spin_op.cpp 
+   runtime/cudaq/algorithms/py_optimizer.cpp 
+   ../runtime/cudaq/platform/common/QuantumExecutionQueue.cpp
+
+  EMBED_CAPI_LINK_LIBS
+   CUDAQuantumMLIRCAPI
+   MLIRCAPIExecutionEngine
+  PRIVATE_LINK_LIBS
+   OptCodeGen
+   OptTransforms
+)
+
+target_include_directories(CUDAQuantumPythonSources.Extension INTERFACE 
+    ${CMAKE_CURRENT_SOURCE_DIR} 
+    ${CMAKE_SOURCE_DIR}/python 
+    ${CMAKE_SOURCE_DIR}/python/utils)
+target_link_libraries(CUDAQuantumPythonSources.Extension INTERFACE 
+    cudaq 
+    cudaq-common 
+    cudaq-em-default
+    fmt::fmt-header-only
+    )
+
+################################################################################
+# Common CAPI
+################################################################################
+
+add_mlir_python_common_capi_library(CUDAQuantumPythonCAPI
+  INSTALL_COMPONENT CUDAQuantumPythonModules
+  INSTALL_DESTINATION mlir_cudaq/_mlir_libs
+  OUTPUT_DIRECTORY "${MLIR_BINARY_DIR}/mlir_cudaq/_mlir_libs"
```

Comment:
I notice that the python bindings here are left in the top-level of the developer's `build/` directory. One thing I've found useful as a developer is to have the python bindings dumped in a stand-alone folder within the build directory, like `build/python/` instead of `build/`. There are few reasons for this - it better emulates the python bindings as a stand-alone package (easy to point PYTHONPATH too), but also for developers switching between this branch and `main` care has to be taken to delete the python bindings in the top-level of `build` before using those in`build/python/`, otherwise you get conflicts. If we stick with the `build/python` for both, then the bindings from one branch are replaced with the bindings for the other branch when you rebuild.

Reply:
will do